hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java

```diff
@@ -0,0 +1,895 @@
+            return;
+        }
+
+        MemberMap memberMap = getMemberMap();
+        MembersView membersView = memberMap.toMembersView();
+
+        for (MemberImpl member : memberMap.getMembers()) {
+            if (member.localMember()) {
+                continue;
+            }
+
+            MembersUpdateOp op = new MembersUpdateOp(member.getUuid(), membersView,
+                    clusterService.getClusterTime(), null, false);
+            op.setCallerUuid(node.getThisUuid());
+            nodeEngine.getOperationService().send(op, member.getAddress());
+        }
+    }
+
+    // handles both new and left members
+    void updateMembers(MembersView membersView) {
+        MemberMap currentMemberMap = memberMapRef.get();
+
+        Collection<MemberImpl> addedMembers = new LinkedList<MemberImpl>();
+        Collection<MemberImpl> removedMembers = new LinkedList<MemberImpl>();
+        ClusterHeartbeatManager clusterHeartbeatManager = clusterService.getClusterHeartbeatManager();
+
+        MemberImpl[] members = new MemberImpl[membersView.size()];
+        int memberIndex = 0;
+        for (MemberInfo memberInfo : membersView.getMembers()) {
+            Address address = memberInfo.getAddress();
+            MemberImpl member = currentMemberMap.getMember(address);
+
+            if (member != null && member.getUuid().equals(memberInfo.getUuid())) {
+                members[memberIndex++] = member;
+                continue;
+            }
+
+            if (member != null) {
+                // uuid changed: means member has gone and come back with a new uuid
+                removedMembers.add(member);
+            }
+
+            member = createMember(memberInfo);
+            addedMembers.add(member);
+
+            long now = clusterService.getClusterTime();
+            clusterHeartbeatManager.onHeartbeat(member, now);
+            clusterHeartbeatManager.acceptMasterConfirmation(member, now);
+
+            clusterService.repairPartitionTableIfReturningMember(member);
+            members[memberIndex++] = member;
+        }
+
+        MemberMap newMemberMap = membersView.toMemberMap();
+        for (MemberImpl member : currentMemberMap.getMembers()) {
+            if (!newMemberMap.contains(member.getAddress())) {
+                removedMembers.add(member);
+            }
+        }
+
+        setMembers(MemberMap.createNew(membersView.getVersion(), members));
+
+        for (MemberImpl member : removedMembers) {
+            handleMemberRemove(memberMapRef.get(), member);
+        }
+
+        sendMembershipEvents(currentMemberMap.getMembers(), addedMembers);
+
+        MemberMap membersRemovedInNotActiveState = membersRemovedInNotActiveStateRef.get();
+        membersRemovedInNotActiveStateRef.set(MemberMap.cloneExcluding(membersRemovedInNotActiveState, members));
+
+        clusterHeartbeatManager.heartbeat();
```

Comment:
The heartbeat method calculates the clock drift based on the last invocation - https://github.com/hazelcast/hazelcast/blob/111a9deb0636fa6d87327f714c8ca2676f592fd6/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java#L387.
So if it called in between periodic invocations, the absolute clock jump could be up to `heartbeatIntervalMillis/2`. I might be wrong here though.